openslides/utils/consumers.py

```diff
@@ -0,0 +1,306 @@
+from typing import Any, Dict, List, Optional
+
+from asgiref.sync import sync_to_async
+from channels.db import database_sync_to_async
+from channels.generic.websocket import AsyncJsonWebsocketConsumer
+
+from ..core.config import config
+from ..core.models import Projector
+from .auth import a_anonymous_is_enabled, has_perm
+from .cache import element_cache, split_element_id
+from .collection import AutoupdateFormat  # noqa
+from .collection import (
+    Collection,
+    CollectionElement,
+    format_for_autoupdate,
+    from_channel_message,
+)
+
+
+class SiteConsumer(AsyncJsonWebsocketConsumer):
+    """
+    Websocket Consumer for the site.
+    """
+    groups = ['site']
+
+    async def connect(self) -> None:
+        """
+        A user connects to the site.
+
+        If it is an anonymous user and anonymous is disabled, the connection is closed.
+
+        Sends the startup data to the user.
+        """
+        # TODO: add a way to ask for the data since a change_id and send only data that is newer
+        if not await a_anonymous_is_enabled() and self.scope['user'].id is None:
+            await self.close()
+        else:
+            await self.accept()
```

Comment:
In channels docs it "is recommended that accept() be called as the last action in connect() if you choose to accept the connection." See https://channels.readthedocs.io/en/latest/tutorial/part_2.html. Is this possible for us?

Reply:
No. The connection has to be accepted, before data can be send. You get ```Socket has not been accepted, so cannot send over it``` when you change the order of the lines.